devito/function.py

```diff
@@ -1025,18 +1022,20 @@ def argument_values(self, alias=None, **kwargs):
         :param kwargs: Dictionary of user-provided argument overrides.
         :param alias: (Optional) name under which to store values.
         """
-        # Take a copy of the replacement before super pops it from kwargs
-
-        new = kwargs.get(self.name)
+        values = {}
         key = alias or self
-
-        if new is not None and isinstance(new, SparseFunction):
-            # If we've been replaced with a SparseFunction,
-            # we need to re-derive defaults and values...
-            values = new.argument_defaults(alias=key).reduce_all()
-        else:
-            # ..., but if not, we simply need to recurse over children.
-            values = self.coordinates.argument_values(alias=key, **kwargs)
+        # Add value override for own data if it is provided
+        if self.name in kwargs:
+            # Update function value
+            values = super(SparseFunction, key).argument_values(alias=key, **kwargs)
+            # Get coordinates if input is SparseFunction else take self
+            new = kwargs.get(self.name, self)
+            new = new if hasattr(new, 'coordinates') else self
```

Comment:
I don't understand why you need to treat the ``else``; if the input were not a ``SparseFunction``, then ``self`` is already handled by ``super``.

I also think you should do ``if isinstance(.., SparseFunction)``, rather than using ``hasattr`` (rule of thumb: stay away from it as much as possible)